### PR TITLE
Let's add images

### PR DESCRIPTION
**Pull request by dfrg**
This patch introduces initial support for rendering images. 

The general process is as follows:
1. Capture image brushes as "patches" at encoding time.
2. During resolve, allocate all images into an atlas, iteratively increasing atlas size on failure until we reach maximum size (currently 8k). Patch draw data stream with atlas locations.
3. At render, create an atlas texture of the necessary dimensions and use `Queue::write_texture` to blit all images to their associated locations.
4. `draw_leaf` computes an inverse transform, same as radial gradients (7 words of draw info).
5. `coarse` writes the image parameters into the ptcl (4 words).
6. `fine` reads the image parameters, converts coords from screen to atlas space, clips, loads 4 pixels and performs bilinear filtering.

I haven't done any performance measurement, but the fine stage hurts a bit. This can be simplified considerably for images where the upper 2x2 matrix is identity and the translation is integral. 

The alternative is to follow the original plan in #176 which calls for rasterizing pre-transformed images into the atlas. This makes fine cheaper at the cost of significantly more complexity at the setup stage and less efficient atlas allocation.

On the todo list: more robustness, retained caching, extend modes, ...

### CODE REVIEW

File: examples/scenes/src/test_scenes.rs
Comment (DJMcNab) [inline examples/scenes/src/test_scenes.rs:8]: Not so sure about using the Piet logo here. Ideally, we'd have a Vello logo, or a Linebender logo.

I don't have a better suggestion for what we can use.

File: examples/scenes/src/test_scenes.rs
Comment (dfrg) [reply examples/scenes/src/test_scenes.rs:106]: This is just my expedient hacky solution to support image loading for examples. Being an example though, it probably does deserve a “don’t do this in real code!” comment.

Since the address is never converted back into a pointer, I don’t think provenance is an issue. Size likely would be on CHERI architectures though.

Comment (dfrg) [reply examples/scenes/src/test_scenes.rs:106]: Fair enough! Pointer to int casts are always worth calling out in code review just to have some discussion logged for future reference.

**Review by armansito — COMMENTED** [review]
Overall looks good. One thing that got me thinking is whether the `ImageCache` should be part of the encoding module or if it should be a separate utility (perhaps part of the scene API). It sounds generally useful and how the atlas is packed is an intrinsic part of the encoding.

When/if we move the encoding module into its own crate, should `ImageCache` be provided by that or should we try to provide a leaner API? I'm not strongly opinionated either way but I am thinking about how to keep the native (C/C++) bridging for clients simple and the library lightweight.

File: examples/scenes/src/test_scenes.rs
Comment (dfrg) [reply examples/scenes/src/test_scenes.rs:8]: I assume we're free to use the druid logo as well but I thought piet might be slightly more closely related to vello given the lineage. For now, the priority is just to have *some* example that displays an image.

I can provide colorful... commentary on a logo project :) I don't think my artistic talents are up to the task!

File: examples/scenes/src/images.rs
Comment (dfrg) [reply examples/scenes/src/images.rs:19]: Didn't realize we already had that dependency. Done!

File: shader/coarse.wgsl
Comment (dfrg) [reply shader/coarse.wgsl:388]: I had to change those values several times while playing with the encoding and missing one of these actually led to a TDR. So yeah, this is supremely annoying.

File: shader/draw_leaf.wgsl
Comment (dfrg) [reply shader/draw_leaf.wgsl:177]: It would be great to do a pass through the code to make use of appropriate matrix types/ops where appropriate. For this PR, I'd like to avoid the extra churn though.

File: src/engine.rs
Comment (dfrg) [reply src/engine.rs:314]: Replaced the magic constant with `format.describe().block_size`. We'll need to revisit if we add support for compressed formats.

Hitting the granularity requirement is actually what led me to using the queue texture write. I'm not sure exactly how the current architecture will be divided for supporting other backends. This is probably a good topic for the next time you crash our meeting :)

**Comment by dfrg** [comment]
> When/if we move the encoding module into its own crate, should `ImageCache` be provided by that or should we try to provide a leaner API? I'm not strongly opinionated either way but I am thinking about how to keep the native (C/C++) bridging for clients simple and the library lightweight.

Doing the final packing requires resolving the atlas layout in some way. This currently sits in the encoding module, but I can easily see breaking that one up into both `encode` and `resolve` with traits for pluggable caching mechanisms (both glyphs and images) on the resolve side. The current state is not very modular, but I think the boundaries are sharp enough to increase the modularity without too much trouble.

File: shader/coarse.wgsl
Comment (armansito) [reply shader/coarse.wgsl:388]: On the naga side this seems to tracked by https://github.com/gfx-rs/naga/issues/2106. I'm putting a mention here and I'll ping the naga issue as well as this would be a huge quality of life improvement (I've wanted this in other projects as well).

**Comment by armansito** [comment]
> > When/if we move the encoding module into its own crate, should `ImageCache` be provided by that or should we try to provide a leaner API? I'm not strongly opinionated either way but I am thinking about how to keep the native (C/C++) bridging for clients simple and the library lightweight.
> 
> Doing the final packing requires resolving the atlas layout in some way. This currently sits in the encoding module, but I can easily see breaking that one up into both `encode` and `resolve` with traits for pluggable caching mechanisms (both glyphs and images) on the resolve side. The current state is not very modular, but I think the boundaries are sharp enough to increase the modularity without too much trouble.

OK, let's discuss this on Monday. I don't think we need to block this PR on this issue but I think let's also plan to talk about what the FFI bridge should look like. Once we pull encoding into its own separate crate we can measure the code size and see if it's worth modularizing.

**Review by armansito — APPROVED** [review]
I would say Raph should probably take a look as well but LGTM from my point of view.

I left two more comments around some locally unused imports. I tested the with_winit/headless examples on M1 Max and with_winit on Android and didn't run into any issues.

As you mentioned, there is a slight performance decrease that I noticed on Pixel 6 (framerate for the scene drops from ~80fps to ~70fps). I didn't see a performance drop on M1 Max.

It might be a good idea to add a test scene with many images just so we get an idea about performance when there are many atlas lookups.

File: shader/draw_leaf.wgsl
Comment (dfrg) [reply shader/draw_leaf.wgsl:177]: sgtm!

File: examples/scenes/src/test_scenes.rs
Comment (dfrg) [reply examples/scenes/src/test_scenes.rs:1]: Fixed!

File: examples/scenes/src/test_scenes.rs
Comment (dfrg) [reply examples/scenes/src/test_scenes.rs:4]: Fixed! There are a few more issues with unused symbols. They've been lingering a while so I'll do another pass to fix those soon (and also make the codebase clippy clean)

**Comment by dfrg** [comment]
> OK, let's discuss this on Monday. I don't think we need to block this PR on this issue but I think let's also plan to talk about what the FFI bridge should look like. Once we pull encoding into its own separate crate we can measure the code size and see if it's worth modularizing.

Looking forward to it!

File: shader/coarse.wgsl
Comment (dfrg) [reply shader/coarse.wgsl:388]: Thanks for the link. This will hopefully make the cut for the next release.

File: shader/fine.wgsl
Comment (raphlinus) [inline shader/fine.wgsl:300]: I'm curious why the by-hand bilinear interpolation here rather than using hardware. Going over the WGSL intrinsics, I think `textureSampleLevel` is the most appropriate version - unlike `textureSample` it doesn't rely on derivatives, and I think it's likely to be more performant than `textureSampleGrad` (which based on past piet-gpu experimentation was slow).

That could perhaps be followup work to this patch.

File: src/encoding/draw.rs
Comment (raphlinus) [inline src/encoding/draw.rs:107]: Would [u16; 2] be a better choice here? It feels more semantic, but perhaps just using u32 leaves less slop around stuff like endianness.

File: shader/fine.wgsl
Comment (raphlinus) [inline shader/fine.wgsl:297]: My gut feeling is that performance might be improved considerably if this is also predicated on `area[i] != 0.0`, though that will almost certainly depend on the hardware. These loads are already conditional, and I think it can reduce texture sample bandwidth for the tile wastage case. Perhaps it's worth doing careful experiments and landing this with the simplest case.

**Review by raphlinus — APPROVED** [review]
Overall this looks good to me. I am not attached to scaling the image in a rasterization pipeline, though my gut feeling is that it would be higher performance in many cases. Ultimately that's a matter for very careful performance analysis, and the goal for this cycle is to get something that works. This approach also avoids the problems of really huge atlas entries when scaling way up.

Thanks, and excited to see this!

File: shader/fine.wgsl
Comment (dfrg) [reply shader/fine.wgsl:300]: I hadn’t intended to do filtering at all in this PR so it was sort of tacked on as an afterthought. I’ll work on hardware sampling as a followup.

File: src/encoding/draw.rs
Comment (dfrg) [reply src/encoding/draw.rs:107]: Yeah, it was simply a desire to keep the shape of the data the same for both CPU and GPU. Also, when we do add descriptor indexing, the xy field (maybe renamed location) can be used as the array index in that mode.

File: shader/fine.wgsl
Comment (dfrg) [reply shader/fine.wgsl:297]: We had discussed this and I agree that since we’re already branching, might as well add the additional check. Will do before merging.

**Comment by renderdender** [comment]
> Overall looks good. One thing that got me thinking is whether the `ImageCache` should be part of the encoding module or if it should be a separate utility (perhaps part of the scene API). It sounds generally useful and how the atlas is packed is an intrinsic part of the encoding.
> 
> When/if we move the encoding module into its own crate, should `ImageCache` be provided by that or should we try to provide a leaner API? I'm not strongly opinionated either way but I am thinking about how to keep the native (C/C++) bridging for clients simple and the library lightweight.

I've been experimenting with SVG parser and realized indeed it would be better to place `ImageCache` to Scene/Encoding, otherwise there will be circular dependencies if you use '`ImageCache`' from `integrations/vello_svg/`.
Just my humble opinion.

**Comment by dfrg** [comment]
Thanks for pointing this out. For higher level integrations like SVG and Lottie that make use of the scene API, the (resolve time) image cache is essentially an internal implementation detail and users shouldn’t have to worry about it.

**Pull request merged**
